### PR TITLE
Define a more specialized registration function

### PR DESCRIPTION
**Pull request by giuseros**
This is to tackle: https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/661 

What I did: 
- I updated the registration functions to only register dialects/passes we actually use
- I updated the `CMakeLists.txt` to reflect the new outlook

TODO:
- Rename `mlir` folder to `rocmlir`
- Search/replace include guards

@krzysz00 , please could you give it a look to see if this aligns with what you had in mind? 

I will hold on the TODOs to avoid making the review too messy

### CODE REVIEW

File: mlir/lib/CAPI/RegisterRocMLIR/CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
-get_property(conversion_libs GLOBAL PROPERTY MLIR_CONVERSION_LIBS)
+get_property(rocmlir_dialect_libs GLOBAL PROPERTY ROCMLIR_DIALECT_LIBS)
+get_property(rocmlir_conversion_libs GLOBAL PROPERTY ROCMLIR_CONVERSION_LIBS)
 add_mlir_public_c_api_library(MLIRCAPIRegisterRocMLIR
   RegisterRocMLIR.cpp
 
   LINK_LIBS PUBLIC
-  ${dialect_libs}
-  ${conversion_libs}
+  ${rocmlir_dialect_libs}
```
Comment (krzysz00) [inline mlir/lib/CAPI/RegisterRocMLIR/CMakeLists.txt:8]: I think this needs a `${rocmlir_conversion_libs}` too?

**Review by krzysz00 — COMMENTED** [review]
Yeah, I think we can save the `mlir/` -> `rocmlir/` change for a different PR.

I think some of the conversion libs got dropped, which could cause issues with stuff like `--convert-tosa-to-rock`.

Have you observed a shorter build on `check-rocmlir-build-only` and on the static library case?

(This should show up in Jenkins)

**Comment by krzysz00** [comment]
While you're here, can you go in to the test depends and remove mentions of `mlir-opt`, `mlir-translate`, and so on? Those will cause all of upstream to be build for tests, which we don't want, and, last I checked, we don't use them in the tests (you may wont to `grep` to be sure)

File: mlir/lib/CAPI/RegisterRocMLIR/CMakeLists.txt (hunk 1 above)
Comment (giuseros) [reply mlir/lib/CAPI/RegisterRocMLIR/CMakeLists.txt:8]: I thought it as well, but I think these get linked in by the dialects? I was not able to see any linking issues without the `conversion_libs` (and also CI seems to run fine)

**Comment by giuseros** [comment]
To me, the time to build (both the library and the compiler) did not seem to improve by a lot (maybe CI will tell us otherwise).

The problem I found is that libraries like `libMLIRNVGPUDialect.{so,a}` still get built. This is because in upstream they include in the global list of dialects and some components (like the `JitRunner`) link against that list. So, even if we reduce the amount of dependencies in our infrastructure, we still depend on components (like the runner) that depend on libraries we will never use.

**Comment by krzysz00** [comment]
How are the things like JitRunner being pulled into `libroccompiler` or the other externally-facing artifacts? (I can agree that `JitRunner` will show up in our tests, but that's nothing much we can do, but it'll still be good to have the refactor)

**Comment by giuseros** [comment]
So the answer is that `RockOps` was depending on `MLIRJitRunner`. I removed it and a new set of dependencies arose. It all seems fine now. If I build the library (or no tests) we are down to 2500 files to build (~25% reduction), which is what we were expecting.

**Comment by krzysz00** [comment]
Might be worth looking for other dependencies on upstream that we don't need

**Comment by giuseros** [comment]
Yeah I agree, I tried to do that. I think/hope it's now down to the minimum

**Review by krzysz00 — COMMENTED** [review]
Re conversion libraries, I think we should swap `add_mlir_conversion_library` for `add_rocmlir_conversion_library`

And I'd suggest going through the C API and removing our dependency on MLIRCAPIRegisterEverything, which might be part of what's pulling in all the conversion libs

**Comment by krzysz00** [comment]
(and I worked out why no linker errors wrt conversion libs: `Pipelines` explicitly lists out its dependencies. We could probably make that slightly less error-prone by using a hypothetical ROCMLIR_CONVERSION_LIBS there)

**Comment by giuseros** [comment]
I removed the `RegisterEverything` calls from the C api, and removed some dependencies from `Pipelines` (and around). I am a bit reluctant to add a fixed list of conversion libraries, because some dialects will link against things they don't really need. While now they are linking against the `MLIRRock*` transforms + additional stuff they really need

**Review by krzysz00 — COMMENTED** [review]
Ok, so, to rephrase my comment. To cleanly handle all our tools (like `rocmlir-gen`, `rocmlir-opt`, etc.) the conversion libraries should be gathered into a variable somewhere that's rocmlir-specific. That could be `ROCMLIR_CONVERSION_LIBS`, or just adding to `ROCMLIR_DIALECT_LIBS`, or something - the point is, they shouldn't be going through `add_mlir_conversion_lib` because that'll leak into upstream.

Also, on the "removing dependencies" front, a big one (esp. because I've gotten burned by big compile times on this before). We don't need `::test::registerTestDialect()` or the like in the `rocmlir-*` tools, and so the test libraries don't need to include MLIRTestIR or MLIRTestPasses.

(I'd also think, for the sake of cleanliness, it might be good idea to have MLIRRockTestPasses add itself to the test libraries, for locality's sake, but that's not a huge thing)

Thanks for doing all the deep dives here!

File: mlir/include/mlir/InitRocMLIRDialects.h [hunk 2]
```diff
@@ -14,21 +14,91 @@
...
 #include "mlir/IR/Dialect.h"
 
 namespace mlir {
 
 // Add all the MLIR dialects to the provided registry.
 inline void registerRocMLIRDialects(DialectRegistry &registry) {
   // clang-format off
-  registry.insert<rock::RockDialect,
-                  migraphx::MIGraphXDialect>();
+  registry.insert<
```
Comment (krzysz00) [inline mlir/include/mlir/InitRocMLIRDialects.h]: Minor comment - could this be factored into a `inline void registerUpstreamDialects()` just so it's easier to think to look there?

**Comment by giuseros** [comment]
Ok, I tried to apply all the changes. Dev machine was down, so I couldn't fully run the tests locally, let's see what CI says :)

**Review by krzysz00 — APPROVED** [review]
LGTM given that everything passes

**Pull request merged**